File: src/mdsite/nav.py
"""Build the nav tree + prev/next ordering from page records."""

from __future__ import annotations

import functools
from dataclasses import dataclass, field
from pathlib import PurePosixPath


@dataclass
class Page:
    rel: str          # source-relative path, forward-slash normalized
    url: str          # clean output URL
    title: str
    order: int | None
    is_index: bool


@dataclass
class NavNode:
    name: str
    url: str | None = None
    title: str | None = None
    children: list["NavNode"] = field(default_factory=list)
    pages: list[Page] = field(default_factory=list)


@dataclass
class Link:
    url: str
    title: str


def _page_sort_key(p: Page):
    # order-present pages first (by order), then by title; mirrors Node.
    has_order = 0 if p.order is not None else 1
    return (has_order, p.order if p.order is not None else 0, p.title.lower())


def _cmp_pages(a: Page, b: Page) -> int:
    ka, kb = _page_sort_key(a), _page_sort_key(b)
    return -1 if ka < kb else (1 if ka > kb else 0)


def build_nav(pages: list[Page]):
    """Return (tree, ordered). tree is the root NavNode; ordered is a flat
    list of Link in nav order for prev/next."""
    # Mutable build phase uses dict children keyed by segment name.
    root_children: dict = {}
    root = {"name": "", "url": None, "title": None, "children": root_children, "pages": []}

    for page in pages:
        segments = page.rel.split("/")
        dir_segments = segments[:-1]
        node = root
        for seg in dir_segments:
            if seg not in node["children"]:
                node["children"][seg] = {
                    "name": seg, "url": None, "title": seg,
                    "children": {}, "pages": [],
                }
            node = node["children"][seg]
        if page.is_index:
            node["url"] = page.url
            node["title"] = page.title
        else:
            node["pages"].append(page)

    def finalize(node) -> NavNode:
        children = sorted(
            (finalize(c) for c in node["children"].values()),
            key=lambda c: c.name.lower(),
        )
        pages = sorted(node["pages"], key=functools.cmp_to_key(_cmp_pages))
        return NavNode(
            name=node["name"], url=node["url"], title=node["title"],
            children=children, pages=pages,
        )

    tree = finalize(root)

    ordered: list[Link] = []

    def flatten(node: NavNode) -> None:
        if node.url:
            ordered.append(Link(url=node.url, title=node.title or ""))
        for page in node.pages:
            ordered.append(Link(url=page.url, title=page.title))
        for child in node.children:
            flatten(child)

    flatten(tree)
    return tree, ordered


def prev_next_map(ordered: list[Link]) -> dict:
    """Map url -> {'prev': Link|None, 'next': Link|None}."""
    result: dict = {}
    for i, link in enumerate(ordered):
        result[link.url] = {
            "prev": ordered[i - 1] if i > 0 else None,
            "next": ordered[i + 1] if i < len(ordered) - 1 else None,
        }
    return result


def is_index_file(rel: str) -> bool:
    base = PurePosixPath(rel.replace("\\", "/")).stem.lower()
    return base in ("index", "readme")

File: src/nav.js
import path from 'node:path';

// Given a flat list of page records, build a nested nav tree mirroring the
// folder structure, plus a flat ordered list for prev/next.
//
// page record: { rel, url, title, order, isIndex }
//   rel     - source-relative path, forward-slash normalized
//   url     - clean output URL
//   title   - resolved page title
//   order   - front-matter order (number) or null
//   isIndex - true for index.md / readme.md (the folder's landing page)
//
// Returns { tree, ordered }.
//   tree    - { name, url, title, children: [], pages: [] } folder nodes
//   ordered - flat array of pages in nav order (for prev/next)

function sortPages(a, b) {
  const ao = a.order, bo = b.order;
  if (ao != null && bo != null && ao !== bo) return ao - bo;
  if (ao != null && bo == null) return -1;
  if (ao == null && bo != null) return 1;
  return a.title.localeCompare(b.title, undefined, { sensitivity: 'base' });
}

function sortFolders(a, b) {
  return a.name.localeCompare(b.name, undefined, { sensitivity: 'base' });
}

export function buildNav(pages) {
  const root = { name: '', url: null, title: null, children: new Map(), pages: [] };

  for (const page of pages) {
    const segments = page.rel.split('/');
    const dirSegments = segments.slice(0, -1);

    let node = root;
    for (const seg of dirSegments) {
      if (!node.children.has(seg)) {
        node.children.set(seg, { name: seg, url: null, title: seg, children: new Map(), pages: [] });
      }
      node = node.children.get(seg);
    }

    if (page.isIndex) {
      // Index page represents the folder itself: set its url/title.
      node.url = page.url;
      node.title = page.title;
    } else {
      node.pages.push(page);
    }
  }

  // Recursively materialize Maps into sorted arrays.
  function finalize(node) {
    const children = [...node.children.values()].map(finalize).sort(sortFolders);
    const pages = node.pages.slice().sort(sortPages);
    return { name: node.name, url: node.url, title: node.title, children, pages };
  }
  const tree = finalize(root);

  // Flatten for prev/next: depth-first, folder landing page first, then its
  // pages, then its subfolders.
  const ordered = [];
  function flatten(node) {
    // Folder landing page (or root home page) comes first.
    if (node.url) {
      ordered.push({ url: node.url, title: node.title });
    }
    for (const page of node.pages) {
      ordered.push({ url: page.url, title: page.title });
    }
    for (const child of node.children) {
      flatten(child);
    }
  }
  flatten(tree);

  return { tree, ordered };
}

// Compute prev/next neighbors for each url from the ordered list.
export function prevNextMap(ordered) {
  const map = new Map();
  for (let i = 0; i < ordered.length; i++) {
    map.set(ordered[i].url, {
      prev: i > 0 ? ordered[i - 1] : null,
      next: i < ordered.length - 1 ? ordered[i + 1] : null,
    });
  }
  return map;
}

// Is this source-relative path an index/landing page for its folder?
export function isIndexFile(rel) {
  const base = path.parse(rel).name.toLowerCase();
  return base === 'index' || base === 'readme';
}
